Comments from Nick Sullivan (#872)
mlswg/mls-protocol: draft-ietf-mls-protocol.md

~~~
{: #remove-flow title="Client Z removes client B from the group"}

Note that the flows in this section are examples; applications can arrange
message flows in other ways.  For example:

* Welcome messages don't necessarily need to be sent directly to new joiners.
  Since they are encrypted to new joiners, they could be distributed more
  broadly, say if the application only had access to a broadcast channel for the
  group.

> * Proposal messages don't necessarily need to be sent to all group members. They
>   only need to reach a member who will commit them.

Author: ```suggestion
* Proposal messages don't need to be immediately sent to all group members.  They need to
  be available to the committer before generating a commit, and to other members before
  processing the commit.
```

Author: @Bren2010 look good?

|              |              +---------------------------->|
|              |              |              | Commit(ExtZ) |
|<----------------------------------------------------------+
|              |<-------------------------------------------+
|              |              |<----------------------------+
|              |              |              |              |
~~~
{: #groupinfo-flow title"Client A publishes a GroupInfo object and Client Z uses
it to join the group"}

## Relationships Between Epochs

A group has a single linear sequence of epochs. Groups and epochs are generally
independent of one another. However, it can sometimes be useful to link epochs
cryptographically, either within a group or across groups. MLS derives a
resumption pre-shared key (PSK) from each epoch to allow entropy extracted from
one epoch to be injected into a future epoch.  A group member that wishes to
inject a PSK issues a PreSharedKey proposal ({{presharedkey}}) describing the
PSK to be injected.  When this proposal is committed, the corresponding PSK will
be incorporated into the key schedule as described in {{pre-shared-keys}}.

> Linking epochs in this way

Reviewer: strange line break

Author: Doesn't change the rendering.  I'll fix it if I end up having to do edits locally, vs in the GitHub UI :)

with `2^d` leaves has `2^(d+1) - 1` nodes, including parent nodes.

Each node in a ratchet tree is either _blank_ (containing no value) or it holds
an HPKE public key with some associated data:

* A public key (for the HPKE scheme in use, see {{ciphersuites}})
* A credential (only for leaf nodes, see {{credentials}})
* An ordered list of "unmerged" leaves (see {{views}})
* A hash of certain information about the node's parent, as of the last time the
  node was changed (see {{parent-hashes}}).

As described in {{views}}, different members know different subsets of the set
of private keys corresponding to the public keys in nodes in the tree.  The
private key corresponding to a parent node is known only to members at leaf
nodes that are descedants of that node.  The private key corresponding to a leaf
node is known only to the member at that leaf node.  A leaf node is _unmerged_

> relative to one of its ancestor node if the member at the leaf node does not

Reviewer 2: ```suggestion
relative to one of its ancestor nodes if the member at the leaf node does not
```

> know the private key corresponding the ancestor node.

Reviewer 2: ```suggestion
know the private key corresponding to the ancestor node.
```

  * If the LeafNode appears in the `leaf_node` value of the UpdatePath in
    a Commit, verify that `leaf_node_source` is set to `commit`.

* Verify that the leaf node's `signature_key` value does not appear in any other
  leaf node in the group's ratchet tree.

* Verify that the leaf node's `encryption_key` value does not appear in any
  other node of the group's ratchet tree (including both leaf nodes and parent

>   nodes).

Reviewer 3: What exactly does it mean to "appear in any other node"? Is it OK e.g. if a signature or an HPKE key of a leaf appears in an intermediate certificate in the X509 chain in the credential of another leaf? Should we also check that the keys don't appear in other places e.g. in the external signers extension or as the HPKE key for external joiners?

them to the Delivery Service.  This way, the key material used to add a member
to a new group is more likely to be fresh and less likely to be compromised.

## Uniqueness of Ratchet Tree Key Pairs

The encryption and signature keys stored in ratchet tree nodes MUST be distinct
from one another.  If two members' leaf nodes have the same signature key, then
the data origin authentication properties afforded by signatures within the
group are degraded.  If two nodes have the same encryption key, it creates a
"double join" scenario where a removed member may still be able to compute the

> group's secrets.

Reviewer 3: I'm not sure I get this argument. First, keys could be related in different ways, e.g. `pk2 = pk1 ^ 2`. If I know `sk1` corresponding to `pk1`, I can decrypt ciphertexts for `sk2 = 2 * sk1` corresponding to `pk2` as well. In general, such relations can't be detected, so I don't see how this check _can_ improve security. On the other hand, it's a bit annoying to implement.

Another reason I don't think this check prevents double joins by malicious insiders. Say leaves A and B have the same HPKE key. This could be because:

* A is honest and a malicious B copied A's key. Since B doesn't know the secret key, this is not an attack and B can be removed.
* A is honest and generated the same key as a malicious B. This clearly can't happen because keys are random.
* A and B are malicious, they collude and they introduced the same key. In this case security is restored only after removing both insiders. (Since A and B collude, they are one entity in a way)
